### PR TITLE
[WIP] Set rate limiter priority dynamically and pass it to FS if needed

### PR DESCRIPTION
**Pull request by gitbw95**
### Context:
Background compactions and flush generate large reads and writes, and can be long running, especially for universal compaction. In some cases, this can impact foreground reads and writes by users. 

From the RocksDB perspective, there can be two kinds of rate limiters, the internal (native) one and the external one. 
- The internal (native) rate limiter is introduced in [the wiki](https://github.com/facebook/rocksdb/wiki/Rate-Limiter). Currently, only IO_LOW and IO_HIGH are used and they are set statically. 
- For the external rate limiter, in FSWritableFile functions,  IOOptions is open for end users to set and get rate_limiter_priority for their own rate limiter. Currently, RocksDB doesn’t pass the rate_limiter_priority through IOOptions to the file system.

### Solution
During the User Read, Flush write, Compaction read/write, the WriteController is used to determine whether DB writes are stalled or slowed down. The rate limiter priority (Env::IOPriority) can be determined accordingly. If the internal (native) rate limiter is not enabled, this priority will be passed to the file system. We would like to set the rate limiter priority at file level, since the Flush/Compaction job level may be too coarse with multiple files and block IO level is too granular.

**TODO:** Add two tables about how the rate limiter priority is determined.

**This PR is for the Write path.** Flush and Compaction writes share the same call path through BlockBaseTableWriter, WritableFileWriter, and FSWritableFile. When a new FSWritableFile object is created, its io_priority_ can be set dynamically based on the state of the WriteController. In WritableFileWriter, before the call sites of FSWritableFile functions, WritableFileWriter::DecideRateLimiterPriority() determines the rate_limiter_priority. If the internal (native) rate limiter is not enabled, the options (IOOptions) argument of FSWritableFile functions will be updated with the rate_limiter_priority. If the internal rate limiter is enabled, the dynamic priority is used rather than the current hardcoded one.

### Test Plan: TBA

### CODE REVIEW

File: file/writable_file_writer.cc [hunk 1]
```diff
@@ -490,19 +500,21 @@ IOStatus WritableFileWriter::WriteBuffered(
...
-    if (rate_limiter_ != nullptr &&
-        rate_limiter_priority_used != Env::IO_TOTAL) {
-      allowed = rate_limiter_->RequestToken(left, 0 /* alignment */,
-                                            rate_limiter_priority_used, stats_,
-                                            RateLimiter::OpType::kWrite);
-    } else {
-      allowed = left;
+    size_t allowed = left;
+    if (rate_limiter_priority_used != Env::IO_TOTAL) {
+      if (rate_limiter_ != nullptr) {
```
Comment (anand1976) [inline file/writable_file_writer.cc:511]: I think we decided to always pass priority in IOOptions. What the file system does with it should be a contract between the user and the fs.

Comment (gitbw95) [reply file/writable_file_writer.cc:511]: @ajkr Please be aware about this final decision.

Comment (gitbw95) [reply file/writable_file_writer.cc:511]: I will update the logic.

File: file/writable_file_writer.cc [hunk 2]
```diff
@@ -211,6 +213,8 @@ IOStatus WritableFileWriter::Close() {
   s = Flush();  // flush cache to OS
 
   IOStatus interim;
+  IOOptions io_options;
+  io_options.rate_limiter_priority = writable_file_->GetIOPriority();
```
[the remark this answers is not in the document]
Comment (gitbw95) [reply file/writable_file_writer.cc:217]: I will leave this into a follow up PR.

File: db/flush_job.cc [hunk 3]
```diff
@@ -1064,6 +1068,21 @@ std::unique_ptr<FlushJobInfo> FlushJob::GetFlushJobInfo() const {
...
+  WriteController* write_controller =
+      versions_->GetColumnFamilySet()->write_controller();
+
+  if (op_type == RateLimiter::OpType::kWrite) {
+    if (write_controller->IsStopped() || write_controller->NeedsDelay()) {
+      return Env::IO_USER;
+    }
+    return Env::IO_HIGH;
+  } else {
+    return Env::IO_USER;
```
[the remark this answers is not in the document]
Comment (gitbw95) [reply db/flush_job.cc:1080]: Will update.

[the remark this answers is not in the document]
Comment (gitbw95) [reply db/flush_job.cc:1080]: @anand1976  https://fburl.com/dapnwp9y indicate the Flush Job read is actually regard as User read. The verification read should be the only Flush read, so I will keep it as is. We can discuss more on this.

[the remark this answers is not in the document]
Comment (anand1976) [reply db/flush_job.cc:1080]: Good point. The default rate_limiter_priority in ReadOptions is IO_TOTAL, which means disable rate limiting. Since this PR is for the write path, we can defer this to the read path PR and discuss it then.

File: file/writable_file_writer.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (hx235) [reply file/writable_file_writer.cc:218]: > Its confusing to have both rate_limiter_priority and prio in IOOptions.

Catching on this proposal - will share my thoughts.

[the remark this answers is not in the document]
Comment (hx235) [reply file/writable_file_writer.cc:218]: Being discussed in internal chat.

File: db/flush_job_test.cc [hunk 4]
```diff
@@ -528,6 +528,72 @@ TEST_F(FlushJobTest, Snapshots) {
...
+  ASSERT_EQ(flush_job.GetRateLimiterPriorityForWrite(), Env::IO_HIGH);
+
+  WriteController* write_controller =
+      flush_job.versions_->GetColumnFamilySet()->write_controller();
+
+  {
+    // When the state from WriteController is Delayed.
+    std::unique_ptr<WriteControllerToken> delay_token =
+        write_controller->GetDelayToken(1000000);
+    ASSERT_EQ(flush_job.GetRateLimiterPriorityForWrite(), Env::IO_USER);
```
Comment (anand1976) [inline db/flush_job_test.cc:586]: Can we somehow also verify that this priority is passed down to the WritableFileWriter/FSWritableFile by the flush job.

File: db/compaction/compaction_job_test.cc [hunk 5]
```diff
@@ -390,6 +391,39 @@ class CompactionJobTestBase : public testing::Test {
...
+      std::unique_ptr<WriteControllerToken> compaction_pressure_token =
+          write_controller->GetCompactionPressureToken();
+      ASSERT_EQ(compaction_job.GetRateLimiterPriorityForWrite(), Env::IO_HIGH);
+    }
+
+    {
+      // When the state from WriteController is Delayed.
+      std::unique_ptr<WriteControllerToken> delay_token =
+          write_controller->GetDelayToken(1000000);
+      ASSERT_EQ(compaction_job.GetRateLimiterPriorityForWrite(), Env::IO_USER);
```
Comment (anand1976) [inline db/compaction/compaction_job_test.cc:418]: Can we verify the priority flows through to the writable file?

**Review by anand1976 — COMMENTED** [review]
LGTM. The calculated rate limiter priority also needs to be used when requesting the rate limiter token. Is there a plan to do that in a subsequent PR?

File: file/writable_file_writer.cc [hunk 6]
```diff
@@ -490,19 +505,19 @@ IOStatus WritableFileWriter::WriteBuffered(
...
   while (left > 0) {
-    size_t allowed;
-    Env::IOPriority rate_limiter_priority_used =
-        WritableFileWriter::DecideRateLimiterPriority(
-            writable_file_->GetIOPriority(), op_rate_limiter_priority);
+    size_t allowed = left;
     if (rate_limiter_ != nullptr &&
         rate_limiter_priority_used != Env::IO_TOTAL) {
       allowed = rate_limiter_->RequestToken(left, 0 /* alignment */,
                                             rate_limiter_priority_used, stats_,
```
Comment (gitbw95) [inline file/writable_file_writer.cc:519]: @anand1976  The new rate_limiter_priority is set for both the internal rate limiter and the downstream IOOptions. Here is one example.

**Comment by gitbw95** [comment]
> 

In this PR, the new rate_limiter_priority is set for both the internal rate limiter and the downstream IOOptions. If there are any missing parts, I will add them in a follow up PR.

File: db/flush_job_test.cc (hunk 4 above)
Comment (gitbw95) [reply db/flush_job_test.cc:586]: This should be doable. I can add this test in a follow up PR.

File: file/writable_file_writer.cc [hunk 7]
```diff
@@ -709,20 +724,20 @@ IOStatus WritableFileWriter::WriteDirect(
...
     // Check how much is allowed
-    size_t size;
-    Env::IOPriority rate_limiter_priority_used =
-        WritableFileWriter::DecideRateLimiterPriority(
-            writable_file_->GetIOPriority(), op_rate_limiter_priority);
+    size_t size = left;
     if (rate_limiter_ != nullptr &&
         rate_limiter_priority_used != Env::IO_TOTAL) {
       size = rate_limiter_->RequestToken(left, buf_.Alignment(),
                                          writable_file_->GetIOPriority(),
```
Comment (anand1976) [inline file/writable_file_writer.cc:739]: One place where the newly calculated rate_limiter_priority is not being used. Another place is line 842.

File: file/writable_file_writer.cc (hunk 6 above)
Comment (anand1976) [reply file/writable_file_writer.cc:519]: There are a couple of places missed - line 739 and 842.

File: util/writable_file_io_priority_test.cc [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+#include <algorithm>
+
+#include "file/writable_file_writer.h"
+#include "rocksdb/file_system.h"
+#include "test_util/testharness.h"
+#include "test_util/testutil.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+class WritableFileWriterIOPriorityTest : public testing::Test {
```
Comment (hx235) [inline util/writable_file_io_priority_test.cc:15]: Should we move this whole test suite into file_reader_writer_test.cc, where there exists `class WritableFileWriterTest : public testing::Test {};`  this test can extend from? This should obsolete this new file along with all the added rule to Target/src.mk/etc in this PR?

File: file/writable_file_writer.cc (hunk 7 above)
Comment (gitbw95) [reply file/writable_file_writer.cc:739]: This one is interesting. In the original code, it is the "writable_file_->GetIOPriority()" rather than "rate_limiter_priority_used", so I kept it as is. I can update it in this PR.

Comment (gitbw95) [reply file/writable_file_writer.cc]: I just updated the code to use rate_limiter_priority_used.

File: util/writable_file_io_priority_test.cc (hunk 8 above)
Comment (gitbw95) [reply util/writable_file_io_priority_test.cc:15]: We have many large files with thousands of lines, so I have tried to split them to smaller ones if possible. WritableFileWriterIOPriorityTest doesn't use existing settings of WritableFileWriterTest, so I added a new test file.

File: db/compaction/compaction_job.cc [hunk 9]
```diff
@@ -2476,6 +2476,21 @@ std::string CompactionJob::GetTableFileName(uint64_t file_number) {
...
+Env::IOPriority CompactionJob::GetRateLimiterPriority() {
+  if (versions_ && versions_->GetColumnFamilySet() &&
+      versions_->GetColumnFamilySet()->write_controller()) {
+    WriteController* write_controller =
+        versions_->GetColumnFamilySet()->write_controller();
+    if (write_controller->NeedsDelay() || write_controller->IsStopped()) {
+      return Env::IO_USER;
+    } else if (write_controller->NeedSpeedupCompaction()) {
+      return Env::IO_HIGH;
+    }
```
Comment (ajkr) [inline db/compaction/compaction_job.cc:2488]: Speedup is not necessarily a problem and already leads to increased compaction parallelism. I am not sure we want to additionally increase I/O priority as it might speed up too much :) (thereby slowing other things down - like memtable flush).

File: file/writable_file_writer.cc [hunk 10]
```diff
@@ -882,16 +899,20 @@ IOStatus WritableFileWriter::WriteDirectWithChecksum(
 #endif  // !ROCKSDB_LITE
 Env::IOPriority WritableFileWriter::DecideRateLimiterPriority(
     Env::IOPriority writable_file_io_priority,
-    Env::IOPriority op_rate_limiter_priority) {
+    Env::IOPriority op_rate_limiter_priority, IOOptions& io_options) {
```
Comment (ajkr) [inline file/writable_file_writer.cc:902]: I think it would be clearer to assign the return value to `IOOptions::rate_limiter_priority` than to mutate the `IOOptions` argument in this function.

File: util/writable_file_io_priority_test.cc (hunk 8 above)
Comment (ajkr) [reply util/writable_file_io_priority_test.cc:15]: I'm OK with test-per-file or test-per-feature,  but not really test-per-(file, function argument). util/file_reader_writer_test.cc seems like the right place for this.

File: file/writable_file_writer.cc (hunk 10 above)
Comment (gitbw95) [reply file/writable_file_writer.cc:902]: Will update.

File: db/compaction/compaction_job.cc (hunk 9 above)
Comment (gitbw95) [reply db/compaction/compaction_job.cc:2488]: Will change it.

**Comment by gitbw95** [comment]
Since there were two objections to writable_file_io_priority_test.cc, I will change it.

File: db/compaction/compaction_job_test.cc (hunk 5 above)
Comment (gitbw95) [reply db/compaction/compaction_job_test.cc]: I can add this test in a follow up PR.

File: util/writable_file_io_priority_test.cc (hunk 8 above)
Comment (gitbw95) [reply util/writable_file_io_priority_test.cc:15]: Since there were two objections to writable_file_io_priority_test.cc, I will change it.

File: HISTORY.md [hunk 11]
```diff
@@ -5,6 +5,7 @@
 * Fixed a bug where RocksDB could corrupt DBs with `avoid_flush_during_recovery == true` by removing valid WALs, leading to `Status::Corruption` with message like "SST file is ahead of WALs" when attempting to reopen.
 * Fixed a bug in async_io path where incorrect length of data is read by FilePrefetchBuffer if data is consumed from two populated buffers and request for more data is sent.
 * Fixed a CompactionFilter bug. Compaction filter used to use `Delete` to remove keys, even if the keys should be removed with `SingleDelete`. Mixing `Delete` and `SingleDelete` may cause undefined behavior.
+* Fixed a bug in `WritableFileWriter::WriteDirect` and `WritableFileWriter::WriteDirectWithChecksum`. The rate_limiter_priority_used was decided but was not used in `rate_limiter_->RequestToken`.
```
Comment (anand1976) [inline HISTORY.md:8]: Suggest rewording the 2nd sentence to "The rate_limiter_priority specified in ReadOptions was not passed to the RateLimiter when requesting a token."

**Review by anand1976 — APPROVED** [review]
LGTM. Thanks for addressing all the comments!

File: HISTORY.md (hunk 11 above)
Comment (gitbw95) [reply HISTORY.md:8]: Will update.

File: file/writable_file_writer.cc (hunk 10 above)
Comment (ajkr) [reply file/writable_file_writer.cc]: Sorry I don't understand the change. It looks like now we pass an `Env::IOPriority&` that gets assigned to the same value that the function returns. Why not just use the return value?

File: util/writable_file_io_priority_test.cc (hunk 8 above)
Comment (ajkr) [reply util/writable_file_io_priority_test.cc]: Am I looking at the wrong version of the code? This looks unchanged since the comment thread was resolved

File: file/writable_file_writer.cc (hunk 10 above)
Comment (gitbw95) [reply file/writable_file_writer.cc]: Good point! I will change it.
